java/src/main/java/com/google/appengine/tools/mapreduce/impl/shardedjob/ShardedJobRunner.java

```diff
@@ -446,18 +447,19 @@ private void runAndUpdateTask(final String jobId, final String taskId, final int
   }
 
   private ShardRetryState<T> handleSliceFailure(ShardedJobStateImpl<T> jobState,
-      IncrementalTaskState<T> taskState, RuntimeException ex, boolean abandon) {
-    if (!(ex instanceof RecoverableException) && !taskState.getTask().allowSliceRetry(abandon)) {
-      return handleShardFailure(jobState, taskState, ex);
-    }
-    int attempts = taskState.incrementAndGetRetryCount();
-    if (attempts > jobState.getSettings().getMaxSliceRetries()){
-      log.log(Level.WARNING, "Slice exceeded its max attempts.");
-      return handleShardFailure(jobState, taskState, ex);
+      IncrementalTaskState<T> taskState, RuntimeException ex, boolean failedDueToAbandonedLock) {
+    if (ex instanceof RecoverableException || taskState.getTask().allowSliceRetry(failedDueToAbandonedLock)) {
+      int attempts = taskState.incrementAndGetRetryCount();
+      if (attempts > jobState.getSettings().getMaxSliceRetries()){
+        log.log(Level.WARNING, "Slice exceeded its max attempts.");
+        return handleShardFailure(jobState, taskState, ex);
+      } else {
+        log.log(Level.INFO, "Slice attempt #" + attempts + " failed. Going to retry.", ex);
+      }
+      return null;
     } else {
-      log.log(Level.INFO, "Slice attempt #" + attempts + " failed. Going to retry.", ex);
+      return handleShardFailure(jobState, taskState, ex);
```

Comment:
messages are going to be confusing if we call `handleShardFailure` in `handleSliceFailure`
What's the purpose?

Reply:
I didn't introduce that - it was how it worked before; i just flipped the the conditions because the conditional expressions were overly complicated, imho.